ENT-5337: Resolve issues in RHSM package

DuckBoss opened the pull request.

DuckBoss commented:
**TODO:** Squash commits after PR is updated based on reviews

jirihnidek reviewed (changes requested):
Thanks for this PR. :+1: I have few requests. It seems now that some FIXME comment were not relevant. Sorry for that. I would also like to know some information from @m-horky, because he created some FIXME comments. Details in comments.

DuckBoss commented on test/rhsm/functional/test_profile.py:
```diff
@@ -49,8 +49,8 @@ def test_get_profile_bad_type(self):
         dummy_pkgs = [
-            Package(name="package1", version="1.0.0", release=1, arch="x86_64"),
-            Package(name="package2", version="2.0.0", release=2, arch="x86_64"),
+            Package(name="package1", version="1.0.0", release="1", arch="x86_64"),
```
Are you sure `release` is also supposed to be an `int`?
The previous comment is referring to `epoch`, and `release` is using strings in `/var/lib/rhsm/cache/profile.json` as well.

jirihnidek replied:
Oh, sorry for confusing you. You are right. Ignore this my comment.

DuckBoss commented on src/rhsm/connection.py:
```diff
@@ -1799,16 +1799,15 @@ def unregisterConsumer(self, consumerId: str) -> None:
         """
         method = "/consumers/%s/certificates" % (self.sanitize(consumer_uuid))
-        if len(serials) > 0:
+        if serials:
```
In this case, `serials` can be an empty list. Is it not better to keep it as it is to check for `None` and zero-length list?

jirihnidek replied:
Yeah, you are right. The list can be empty... Both `if serials` or `if serials is not None and len(serials) > 0` would OK.

DuckBoss commented on src/rhsm/connection.py:
```diff
@@ -1501,7 +1500,7 @@ def registerConsumer(
             "type": consumer_type,
             "name": name,
-            "facts": facts,
+            "facts": facts or {},
```
Good point, should be resolved now.

DuckBoss commented on src/rhsm/config.py:
```diff
@@ -143,8 +143,8 @@ def read(self, file_names: Optional[List[str]] = None) -> List[str]:
-        # FIXME Instead of optional list, we can set default to [] instead
-        if file_names is None:
+        file_names: list = file_names or []
+        if not file_names:
```
I did not understand the reasoning behind the FIXME either, but as a compromise I added `line 146` to implement a default value without putting a mutable type in the function arguments.

DuckBoss commented on src/rhsm/config.py:
```diff
@@ -341,8 +341,7 @@ class RhsmHostConfigParser(RhsmConfigParser):
     def __init__(self, config_file: Optional[str] = None, defaults=None) -> None:
-        # FIXME Use super() instead
-        RhsmConfigParser.__init__(self, config_file, defaults)
+        super(RhsmHostConfigParser, self).__init__(config_file, defaults)
```
Thanks, resolved now.

DuckBoss commented on src/rhsm/connection.py:
```diff
@@ -1866,26 +1866,24 @@ def bind(self, consumerId: str, entitle_date: datetime.datetime = None) -> List[
         """
         method = "/consumers/%s/certificates/%s" % (self.sanitize(consumerId), self.sanitize(str(serial)))
-        return self.conn.request_delete(method, description=_("Unsubscribing"))
+        return self.conn.request_delete(method, description=_("Unsubscribing")) is not None
```
I need to check this again. I made this change before removing Restlib so this may no longer be the case where `None` is returned. This may also be a problem for this commit: 7bb6747ea since that was also changed before removing Restlib.

DuckBoss commented on src/rhsm/profile.py:
```diff
@@ -302,16 +302,14 @@ class Package(object):
     Represents a package installed on the system.
     """
 
-    # FIXME: epoch should be probably string (not int)
```
Thanks for the clarification, I have changed it to use `int`.

DuckBoss commented on src/rhsm/connection.py:
```diff
@@ -1799,16 +1799,15 @@ def unregisterConsumer(self, consumerId: str) -> None:
 
-    # FIXME: default value of serials should not be list
-    def getCertificates(self, consumer_uuid: str, serials: list = []) -> List[dict]:
+    def getCertificates(self, consumer_uuid: str, serials: list = None) -> List[dict]:
```
Thanks, resolved now.

DuckBoss commented on src/rhsm/connection.py:
```diff
@@ -1866,26 +1866,24 @@ def bind(self, consumerId: str, entitle_date: datetime.datetime = None) -> List[
         """
         method = "/consumers/%s/entitlements/pool/%s" % (self.sanitize(consumer_uuid), self.sanitize(pool_id))
-        return self.conn.request_delete(method, description=_("Unsubscribing"))
+        return self.conn.request_delete(method, description=_("Unsubscribing")) is not None
```
^ Please see previous comment, I made this change before removing Restlib so I need to revisit this since returning `None` may no longer be the case. Thanks for bringing it to my attention :)

DuckBoss replied:
^^ see previous comment for updates.

jirihnidek commented on src/rhsm/connection.py:
```diff
@@ -2097,14 +2096,13 @@ def activateMachine(self, consumerId: str, email: str = None, lang: str = None)
-            # FIXME: sanitize lang
-            if lang:
-                method += "&email_locale=%s" % lang
+        # FIXME: sanitize email
```
NP, GitHub does not help too much, when PR contains more commit changing same file.

DuckBoss commented on src/rhsm/utils.py:
```diff
@@ -280,6 +280,9 @@ def fix_no_proxy() -> None:
+        stdout: TextIO = None
+        stderr: TextIO = None
         try:
             devnull = open(os.devnull, "w")
```
I don't think I understand what you mean regarding moving those lines. 
Would it instead be better to check if devnull is not None before assigning it to sys.stdout/sys.stderr in case open() raises an exception?
(like this: https://github.com/candlepin/subscription-manager/pull/3168/commits/96d7e0fc277fe887a3c9648edb6c705834e779af)

jirihnidek replied:
It is actually not necessary, because the finally block contains `if stdout is not None` check before restoring `sys.stdout`, etc. I probably overlooked this.

DuckBoss replied:
Thanks for the clarification

jirihnidek commented on src/rhsm/connection.py:
```diff
@@ -1141,7 +1137,15 @@ def _request(
+        except json.JSONDecodeError:
+            # This is primarily intended for getting releases from CDN, because
+            # the file containing releases is plaintext and not json.
+            return result["content"]
```
I guess that we should not solve removing of `RestLib` class using this way and return only content. At least doc string and type-hints do not correspond to the state of code.

Maybe it would be better to return dictionary with result and return `result["content"]` from methods like `request_get()`, `request_post()`... Maybe we should return `result["content"]` later and we should return content from callers of `request_get()`... like from methods of `UEPConnection` class e.g. `getConsumer()`.

It seems that you assume that dictionary `result` is returned. Look at method `unregisterConsumer()`, but it is not true ATM.

DuckBoss replied:
Yes, that would be a better approach. I'll check it out. Thanks for your feedback.

jirihnidek commented on src/rhsm/connection.py:
```diff
@@ -187,10 +187,9 @@ class ConnectionType(enum.Enum):
 
 # FIXME: this concept with reslib_class is horrible. The _request() method can return anything.
 # It does not have much benefit and it is easy to introduce new bug.
```
You removed the restlib_class argument. :+1: Thus this FIXME comment should be removed too.

jirihnidek commented on src/rhsm/connection.py:
```diff
@@ -1860,39 +1744,27 @@ def getOwnerList(self, username: str) -> List[dict]:
+        if result is not None:
+            status = result.get("status")
+            return status == 204
+        return False
```
It looks better, but it actually doesn't work, because `request_delete()` return only content. See my previous comment.

jirihnidek reviewed (changes requested):
Thanks for updates :+1: I found some issues that have to be resolved (more details in comments). We can discuss best solution next week.

jirihnidek commented on src/rhsm/config.py:
```diff
@@ -143,32 +144,35 @@ def read(self, file_names: Optional[List[str]] = None) -> List[str]:
-        # FIXME Stop using tempfile like that
+        # Create a temporary file to write config data to it and
+        # rename the file to the expected config file name after successfully
+        # writing all config data.
```
So far, so good. Next time it is ok to add link of the BZ bug report to the comment of code.

jirihnidek reviewed (changes requested):
Hi, thanks for update. It is better. :+1: I have only few requests.

DuckBoss commented:
Changes since last review:
 - **Rebased branch to catch up with main**
  - Added link to BZ in comments as additional reference
  for using named temporary files for saving config.
  - Corrected docstrings that had old parameter names.
  - Added docstring to new request content extraction method.
  - Renamed validateResponse method to validateResult.
  - Renamed '_request_response_handling' method
  to '_extract_content_from_response'.
  - Changed base exceptions to raise non-generic exceptions.
  - Fixed return behavior for 'unregisterConsumer' method.
  - Renamed 'Restlib...' unit test classes to 'BaseRestLib...'.
  - Added unit test for '_extract_content_from_response' method.

jirihnidek reviewed (approved):
LGTM, Thanks for all updates.
